server/engine/telemetry.go

```diff
@@ -41,29 +57,73 @@ func StoreMetricFromMainOutputs(db *gorm.DB) {
 	db.Where("module_name = ?", "").Find(&mainOutput)
 
 	if loc, exists := mainOutput.Values["location"]; exists {
-		model.SetMetric(db, model.Region, loc.(map[string]interface{})["value"].(string))
+		model.SetMetric(db, model.Region, loc.(map[string]interface{})["value"].(string), "")
 	} else {
 		log.Error("Location of deployment is missing : will not be included in telemetry")
 	}
 	if access, exists := mainOutput.Values["access"]; exists {
-		model.SetMetric(db, model.AccessType, access.(map[string]interface{})["value"].(string))
+		model.SetMetric(db, model.AccessType, access.(map[string]interface{})["value"].(string), "")
 	} else {
 		log.Error("Access Type of deployment is missing : will not be included in telemetry")
 	}
 }
 
+func StoreRetriesPerStep(db *gorm.DB, step model.Step) {
+
+	// len(step.Executions) includes the 1st attempt as well
+	// which should not be considered as a "retry"
+	// In cases where a step is not executed at all, retries should be set to 0.
+	retries := len(step.Executions) - 1
+	if retries < 0 {
+		retries = 0
+	}
+	model.SetMetric(db, model.Retries, fmt.Sprint(retries), step.Name)
+
+}
+
+func StoreErrorsPerStep(db *gorm.DB, step model.Step) {
+
+	// for errors, loop through all executions and combine all errors found for each step and then merge them all
+	for _, execution := range step.Executions {
+		if execution.ErrorDetails != "" {
+			model.SetMetric(db, model.Errors, execution.ErrorDetails, step.Name)
+		}
+	}
+}
+
+func StoreMetricsPerStep(db *gorm.DB) {
```

Comment:
And here as well, does this function need to be exported(start with capital letter)? The only use is in `SetFinalMetrics()`.

Reply:
Updated all function names and made them internal helpers